backend/logger/logger.py

```diff
@@ -0,0 +1,92 @@
+from fileHandler import FileHandler
+from singleton import Singleton
+
+from datetime import datetime
+import os
+from inspect import stack, getframeinfo
+
+
+class Logger(metaclass=Singleton):
+    """
+    Main class for logging functionalities.
+    """
+
+    def __init__(self, file_name="session", level=0):
+        self.level = level
+        fileHandler = FileHandler(file_name)
+        self.writer = fileHandler.getWriteHandle()
+
+    def debug(self, message):
+        """Write a debug message to the file using the writer
+
+        Parameters:
+        message (str) : The message you want to output.
+        """
+        if self.level > 0:
+            return
+        frameinfo = getframeinfo(stack()[1][0])
+        self.writer.write(
```

Comment:
I notice some duplicate code here for the debug, warn, error messages when you have the `self.writer.write` call...could we separate the `self.writer.write` call into a separate function which e.g. takes the log level, frame info and message and just calls the `self.writer.write` part; then all of `debug`, `warn` and `error` functions can call it

Reply:
Yes, that is possible and definitely better stylistically . However it's a bit worrying to me in terms of time efficiency. We aren't planning to add more functions like these so there won't be more duplicate code later than there is now. Making all these functions wrappers to a common writer function can waste a lot of time. For example think about the time it takes to unload the register values of Debug onto the stack, load the Write values into registers, pull the image of Write into memory, do the calculations, then unload the image then return to Debug on the stack and fill the registers again. Do you think it's worth the overhead?

Reply:
Hmm I didn't consider that part, ok let's leave the duplication then?

Reply:
Well I don't know. That's the reason I wrote them that way, if you think the overhead is acceptable I will change it :)

Reply:
What you can do is just define a string in this class like: log_message = "({date} {level}:    {filename}:{line_number}:{message})\n". Then, in the debug/warn/error function, it would just be self.writer.write(log_message.format(date=xxx, level=xxx, filename=xxx, line_number=xxx, message=xxx)).